### ISSUE TITLE
Event Driven Service Discovery: large updates cause BIG-IP to become unstable

### ISSUE THREAD

**Issue by chen23** [F5Networks/f5-appsvcs-extension#308]
### Environment
 * Application Services Version: 3.21
 * BIG-IP Version: 15.1.0.4

### Summary
If you send a large number of nodes to event driven service discovery it will cause the BIG-IP to become unstable.

### Steps To Reproduce
1. Submit the following declaration:
```json
{
    "schemaVersion": "3.9.0",
    "class": "ADC",
    "Tenant": {
        "class": "Tenant",
        "Application": {
            "class": "Application",
            "template": "generic",
            "pool": {
                "class": "Pool",
                "monitors": ["http"],
                "members": [
                    {
                        "addressDiscovery": "event",
                        "servicePort": 80
                    }
                ]
            }
        }
    }
}
```

2. Next send a large payload of nodes.
```json
[{"id":"newNode1","ip":"192.0.2.3"},{"id":"newNode2","ip":"192.0.2.4"},{"id":"newNode5","ip":"192.0.2.5"},{"id":"newNode6","ip":"192.0.2.6"},{"id":"newNode7","ip":"192.0.2.7"},{"id":"newNode8","ip":"192.0.2.8"},{"id":"newNode9","ip":"192.0.2.9"},{"id":"newNode10","ip":"192.0.2.10"},{"id":"newNode11","ip":"192.0.2.11"},{"id":"newNode12","ip":"192.0.2.12"},{"id":"newNode13","ip":"192.0.2.13"},{"id":"newNode14","ip":"192.0.2.14"},{"id":"newNode15","ip":"192.0.2.15"},{"id":"newNode16","ip":"192.0.2.16"},{"id":"newNode17","ip":"192.0.2.17"},{"id":"newNode18","ip":"192.0.2.18"},{"id":"newNode19","ip":"192.0.2.19"},{"id":"newNode20","ip":"192.0.2.20"},{"id":"newNode21","ip":"192.0.2.21"},{"id":"newNode22","ip":"192.0.2.22"},{"id":"newNode23","ip":"192.0.2.23"},{"id":"newNode24","ip":"192.0.2.24"},{"id":"newNode25","ip":"192.0.2.25"},{"id":"newNode26","ip":"192.0.2.26"},{"id":"newNode27","ip":"192.0.2.27"},{"id":"newNode28","ip":"192.0.2.28"},{"id":"newNode29","ip":"192.0.2.29"},{"id":"newNode30","ip":"192.0.2.30"},{"id":"newNode31","ip":"192.0.2.31"},{"id":"newNode32","ip":"192.0.2.32"},{"id":"newNode33","ip":"192.0.2.33"},{"id":"newNode34","ip":"192.0.2.34"},{"id":"newNode35","ip":"192.0.2.35"},{"id":"newNode36","ip":"192.0.2.36"},{"id":"newNode37","ip":"192.0.2.37"},{"id":"newNode38","ip":"192.0.2.38"},{"id":"newNode39","ip":"192.0.2.39"},{"id":"newNode40","ip":"192.0.2.40"},{"id":"newNode41","ip":"192.0.2.41"},{"id":"newNode42","ip":"192.0.2.42"},{"id":"newNode43","ip":"192.0.2.43"},{"id":"newNode44","ip":"192.0.2.44"},{"id":"newNode45","ip":"192.0.2.45"},{"id":"newNode46","ip":"192.0.2.46"},{"id":"newNode47","ip":"192.0.2.47"},{"id":"newNode48","ip":"192.0.2.48"},{"id":"newNode49","ip":"192.0.2.49"},{"id":"newNode50","ip":"192.0.2.50"},{"id":"newNode51","ip":"192.0.2.51"},{"id":"newNode52","ip":"192.0.2.52"},{"id":"newNode53","ip":"192.0.2.53"},{"id":"newNode54","ip":"192.0.2.54"},{"id":"newNode55","ip":"192.0.2.55"},{"id":"newNode56","ip":"192.0.2.56"},{"id":"newNode57","ip":"192.0.2.57"},{"id":"newNode58","ip":"192.0.2.58"},{"id":"newNode59","ip":"192.0.2.59"},{"id":"newNode60","ip":"192.0.2.60"},{"id":"newNode61","ip":"192.0.2.61"},{"id":"newNode62","ip":"192.0.2.62"},{"id":"newNode63","ip":"192.0.2.63"},{"id":"newNode64","ip":"192.0.2.64"},{"id":"newNode65","ip":"192.0.2.65"},{"id":"newNode66","ip":"192.0.2.66"},{"id":"newNode67","ip":"192.0.2.67"},{"id":"newNode68","ip":"192.0.2.68"},{"id":"newNode69","ip":"192.0.2.69"},{"id":"newNode70","ip":"192.0.2.70"},{"id":"newNode71","ip":"192.0.2.71"},{"id":"newNode72","ip":"192.0.2.72"},{"id":"newNode73","ip":"192.0.2.73"},{"id":"newNode74","ip":"192.0.2.74"},{"id":"newNode75","ip":"192.0.2.75"},{"id":"newNode76","ip":"192.0.2.76"},{"id":"newNode77","ip":"192.0.2.77"},{"id":"newNode78","ip":"192.0.2.78"},{"id":"newNode79","ip":"192.0.2.79"},{"id":"newNode80","ip":"192.0.2.80"},{"id":"newNode81","ip":"192.0.2.81"},{"id":"newNode82","ip":"192.0.2.82"},{"id":"newNode83","ip":"192.0.2.83"},{"id":"newNode84","ip":"192.0.2.84"},{"id":"newNode85","ip":"192.0.2.85"},{"id":"newNode86","ip":"192.0.2.86"},{"id":"newNode87","ip":"192.0.2.87"},{"id":"newNode88","ip":"192.0.2.88"},{"id":"newNode89","ip":"192.0.2.89"},{"id":"newNode90","ip":"192.0.2.90"},{"id":"newNode91","ip":"192.0.2.91"},{"id":"newNode92","ip":"192.0.2.92"},{"id":"newNode93","ip":"192.0.2.93"},{"id":"newNode94","ip":"192.0.2.94"},{"id":"newNode95","ip":"192.0.2.95"},{"id":"newNode96","ip":"192.0.2.96"},{"id":"newNode97","ip":"192.0.2.97"},{"id":"newNode98","ip":"192.0.2.98"},{"id":"newNode99","ip":"192.0.2.99"},{"id":"newNode100","ip":"192.0.2.100"},{"id":"newNode101","ip":"192.0.2.101"},{"id":"newNode102","ip":"192.0.2.102"},{"id":"newNode103","ip":"192.0.2.103"},{"id":"newNode104","ip":"192.0.2.104"},{"id":"newNode105","ip":"192.0.2.105"},{"id":"newNode106","ip":"192.0.2.106"},{"id":"newNode107","ip":"192.0.2.107"},{"id":"newNode108","ip":"192.0.2.108"},{"id":"newNode109","ip":"192.0.2.109"},{"id":"newNode110","ip":"192.0.2.110"},{"id":"newNode111","ip":"192.0.2.111"},{"id":"newNode112","ip":"192.0.2.112"},{"id":"newNode113","ip":"192.0.2.113"},{"id":"newNode114","ip":"192.0.2.114"},{"id":"newNode115","ip":"192.0.2.115"},{"id":"newNode116","ip":"192.0.2.116"},{"id":"newNode117","ip":"192.0.2.117"},{"id":"newNode118","ip":"192.0.2.118"},{"id":"newNode119","ip":"192.0.2.119"},{"id":"newNode120","ip":"192.0.2.120"},{"id":"newNode121","ip":"192.0.2.121"},{"id":"newNode122","ip":"192.0.2.122"},{"id":"newNode123","ip":"192.0.2.123"},{"id":"newNode124","ip":"192.0.2.124"},{"id":"newNode125","ip":"192.0.2.125"},{"id":"newNode126","ip":"192.0.2.126"},{"id":"newNode127","ip":"192.0.2.127"},{"id":"newNode128","ip":"192.0.2.128"},{"id":"newNode129","ip":"192.0.2.129"},{"id":"newNode130","ip":"192.0.2.130"},{"id":"newNode131","ip":"192.0.2.131"},{"id":"newNode132","ip":"192.0.2.132"},{"id":"newNode133","ip":"192.0.2.133"},{"id":"newNode134","ip":"192.0.2.134"},{"id":"newNode135","ip":"192.0.2.135"},{"id":"newNode136","ip":"192.0.2.136"},{"id":"newNode137","ip":"192.0.2.137"},{"id":"newNode138","ip":"192.0.2.138"},{"id":"newNode139","ip":"192.0.2.139"},{"id":"newNode140","ip":"192.0.2.140"},{"id":"newNode141","ip":"192.0.2.141"},{"id":"newNode142","ip":"192.0.2.142"},{"id":"newNode143","ip":"192.0.2.143"},{"id":"newNode144","ip":"192.0.2.144"},{"id":"newNode145","ip":"192.0.2.145"},{"id":"newNode146","ip":"192.0.2.146"},{"id":"newNode147","ip":"192.0.2.147"},{"id":"newNode148","ip":"192.0.2.148"},{"id":"newNode149","ip":"192.0.2.149"},{"id":"newNode150","ip":"192.0.2.150"},{"id":"newNode151","ip":"192.0.2.151"},{"id":"newNode152","ip":"192.0.2.152"},{"id":"newNode153","ip":"192.0.2.153"},{"id":"newNode154","ip":"192.0.2.154"},{"id":"newNode155","ip":"192.0.2.155"},{"id":"newNode156","ip":"192.0.2.156"},{"id":"newNode157","ip":"192.0.2.157"},{"id":"newNode158","ip":"192.0.2.158"},{"id":"newNode159","ip":"192.0.2.159"},{"id":"newNode160","ip":"192.0.2.160"},{"id":"newNode161","ip":"192.0.2.161"},{"id":"newNode162","ip":"192.0.2.162"},{"id":"newNode163","ip":"192.0.2.163"},{"id":"newNode164","ip":"192.0.2.164"},{"id":"newNode165","ip":"192.0.2.165"},{"id":"newNode166","ip":"192.0.2.166"},{"id":"newNode167","ip":"192.0.2.167"},{"id":"newNode168","ip":"192.0.2.168"},{"id":"newNode169","ip":"192.0.2.169"},{"id":"newNode170","ip":"192.0.2.170"},{"id":"newNode171","ip":"192.0.2.171"},{"id":"newNode172","ip":"192.0.2.172"},{"id":"newNode173","ip":"192.0.2.173"},{"id":"newNode174","ip":"192.0.2.174"},{"id":"newNode175","ip":"192.0.2.175"},{"id":"newNode176","ip":"192.0.2.176"},{"id":"newNode177","ip":"192.0.2.177"},{"id":"newNode178","ip":"192.0.2.178"},{"id":"newNode179","ip":"192.0.2.179"},{"id":"newNode180","ip":"192.0.2.180"},{"id":"newNode181","ip":"192.0.2.181"},{"id":"newNode182","ip":"192.0.2.182"},{"id":"newNode183","ip":"192.0.2.183"},{"id":"newNode184","ip":"192.0.2.184"},{"id":"newNode185","ip":"192.0.2.185"},{"id":"newNode186","ip":"192.0.2.186"},{"id":"newNode187","ip":"192.0.2.187"},{"id":"newNode188","ip":"192.0.2.188"},{"id":"newNode189","ip":"192.0.2.189"},{"id":"newNode190","ip":"192.0.2.190"},{"id":"newNode191","ip":"192.0.2.191"},{"id":"newNode192","ip":"192.0.2.192"},{"id":"newNode193","ip":"192.0.2.193"},{"id":"newNode194","ip":"192.0.2.194"},{"id":"newNode195","ip":"192.0.2.195"},{"id":"newNode196","ip":"192.0.2.196"},{"id":"newNode197","ip":"192.0.2.197"},{"id":"newNode198","ip":"192.0.2.198"},{"id":"newNode199","ip":"192.0.2.199"},{"id":"newNode200","ip":"192.0.2.200"},{"id":"newNode201","ip":"192.0.2.201"},{"id":"newNode202","ip":"192.0.2.202"},{"id":"newNode203","ip":"192.0.2.203"},{"id":"newNode204","ip":"192.0.2.204"},{"id":"newNode205","ip":"192.0.2.205"},{"id":"newNode206","ip":"192.0.2.206"},{"id":"newNode207","ip":"192.0.2.207"},{"id":"newNode208","ip":"192.0.2.208"},{"id":"newNode209","ip":"192.0.2.209"},{"id":"newNode210","ip":"192.0.2.210"},{"id":"newNode211","ip":"192.0.2.211"},{"id":"newNode212","ip":"192.0.2.212"},{"id":"newNode213","ip":"192.0.2.213"},{"id":"newNode214","ip":"192.0.2.214"},{"id":"newNode215","ip":"192.0.2.215"},{"id":"newNode216","ip":"192.0.2.216"},{"id":"newNode217","ip":"192.0.2.217"},{"id":"newNode218","ip":"192.0.2.218"},{"id":"newNode219","ip":"192.0.2.219"},{"id":"newNode220","ip":"192.0.2.220"},{"id":"newNode221","ip":"192.0.2.221"},{"id":"newNode222","ip":"192.0.2.222"},{"id":"newNode223","ip":"192.0.2.223"},{"id":"newNode224","ip":"192.0.2.224"},{"id":"newNode225","ip":"192.0.2.225"},{"id":"newNode226","ip":"192.0.2.226"},{"id":"newNode227","ip":"192.0.2.227"},{"id":"newNode228","ip":"192.0.2.228"},{"id":"newNode229","ip":"192.0.2.229"},{"id":"newNode230","ip":"192.0.2.230"},{"id":"newNode231","ip":"192.0.2.231"},{"id":"newNode232","ip":"192.0.2.232"},{"id":"newNode233","ip":"192.0.2.233"},{"id":"newNode234","ip":"192.0.2.234"},{"id":"newNode235","ip":"192.0.2.235"},{"id":"newNode236","ip":"192.0.2.236"},{"id":"newNode237","ip":"192.0.2.237"},{"id":"newNode238","ip":"192.0.2.238"},{"id":"newNode239","ip":"192.0.2.239"},{"id":"newNode240","ip":"192.0.2.240"},{"id":"newNode241","ip":"192.0.2.241"},{"id":"newNode242","ip":"192.0.2.242"},{"id":"newNode243","ip":"192.0.2.243"},{"id":"newNode244","ip":"192.0.2.244"},{"id":"newNode245","ip":"192.0.2.245"},{"id":"newNode246","ip":"192.0.2.246"},{"id":"newNode247","ip":"192.0.2.247"},{"id":"newNode248","ip":"192.0.2.248"},{"id":"newNode249","ip":"192.0.2.249"},{"id":"newNode250","ip":"192.0.2.250"},{"id":"newNode251","ip":"192.0.2.251"},{"id":"newNode252","ip":"192.0.2.252"},{"id":"newNode253","ip":"192.0.2.253"},{"id":"newNode254","ip":"192.0.2.254"}]
```

3. Observe the system load increase
```
[admin@ip-10-1-1-4:INOPERATIVE:Changes Pending] ~ # uptime
 05:29:05 up  1:31,  1 user,  load average: 103.94, 72.83, 42.36
```
4. observe a large number of tmsh commands running
```
restnod+ 19286  1153  0 05:16 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode120 address 192.0.2.120 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19291  1153  0 05:16 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode121 address 192.0.2.121 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19413  1153  0 05:16 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode122 address 192.0.2.122 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19566  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode123 address 192.0.2.123 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19831  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode124 address 192.0.2.124 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19848  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode125 address 192.0.2.125 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19869  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode126 address 192.0.2.126 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19945  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode127 address 192.0.2.127 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19967  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode128 address 192.0.2.128 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19968  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode129 address 192.0.2.129 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19970  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode130 address 192.0.2.130 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19971  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode131 address 192.0.2.131 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19972  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode132 address 192.0.2.132 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19978  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode133 address 192.0.2.133 metadata replace-all-with { appsvcs-discovery { } }
restnod+ 19979  1153  0 05:17 ?        S      0:00 tmsh -a create ltm node /Tenant/newNode134 address 192.0.2.134 metadata replace-all-with { appsvcs-discovery { } }
```
### Expected Behavior
Ability to make a large number of updates
### Actual Behavior
Device becomes unresponsive

**Comment by chen23** [comment]
following is a PoC of using a tmsh transaction.  Much, much better.  Note PoC code only (causes some issues where does not delete cleanly, also only optimizes adding nodes).  took about 3 seconds to run with tmsh transaction versus previously did not run to completion.

```
real    0m3.269s
user    0m0.056s
sys     0m0.000s
```

code changes /var/config/rest/iapps/f5-service-discovery/lib

```
*** TmshUtil.js.orig	2020-07-27 17:35:48.811925010 +0000
--- TmshUtil.js	2020-07-29 02:07:51.435747506 +0000
***************
*** 8,16 ****

  'use strict';

  const childProcess = require('child_process');
!
  function generateListCommand(path) {
      return `tmsh -a list ${path}`;
  }

--- 8,16 ----

  'use strict';

  const childProcess = require('child_process');
! const fs = require('fs');
  function generateListCommand(path) {
      return `tmsh -a list ${path}`;
  }

***************
*** 35,42 ****
--- 35,64 ----
  function generateAddNodeCommand(node) {
      // eslint-disable-next-line max-len
      return `tmsh -a create ltm node ${node.id} address ${node.ip} metadata replace-all-with { appsvcs-discovery { } }`;
  }
+ function generateBulkAddNodeCommand(nodes) {
+     // eslint-disable-next-line max-len
+     //
+     let cmd = `cli script service-discovery {
+ proc script::init {} {
+ }
+
+ proc script::run {} {
+   tmsh::begin_transaction\n`;
+     cmd +=  nodes.map((node) => `tmsh::create ltm node ${node.id} address ${node.ip} metadata replace-all-with \\{ appsvcs-discovery \\{ \\} \\}`).join("\n");
+     cmd += `\ntmsh::commit_transaction
+ }
+
+ proc script::help {} {
+ }
+
+ proc script::tabc {} {
+ }
+ }`;
+     return cmd;
+ }
+

  function generateDeleteNodeCommand(path) {
      return `tmsh -a delete ltm node ${path}`;
  }
***************
*** 84,91 ****
--- 106,135 ----
          });
      });
  }

+ function executeMultiLineCommand(command) {
+     return new Promise((resolve, reject) => {
+ //fs = require('fs');
+ fs.writeFile('/var/tmp/service-discovery.cli', command, function (err) {
+   if(err)
+     reject(err);
+ });
+         const tmsh = childProcess.spawn('/bin/tmsh',['-a','-c','load sys config merge file /var/tmp/service-discovery.cli; run cli script service-discovery']);
+         let output = '';
+         tmsh.stdout.on('data', (data) => {
+           output += data;
+         });
+         tmsh.stderr.on('data',(data) => {
+           reject(`${data}`);
+         });
+         //tmsh.stdin.write(command);
+         //tmsh.stdin.end();
+         tmsh.on('close', (code) => {
+           resolve(output);
+           });
+     });
+ }
  function list(path) {
      return executeCommand(generateListCommand(path));
  }

***************
*** 95,102 ****
--- 139,149 ----

  function addNode(node) {
      return executeCommand(generateAddNodeCommand(node));
  }
+ function addBulkNodes(nodes) {
+     return executeMultiLineCommand(generateBulkAddNodeCommand(nodes));
+ }

  function deleteNode(node) {
      return executeCommand(generateDeleteNodeCommand(node));
  }
***************
*** 236,243 ****
--- 283,291 ----
      executeCommand,
      list,
      updatePoolMembers,
      addNode,
+     addBulkNodes,
      deleteNode,
      addFolder,
      addDataGroup,
      updateDataGroup,
```

```
*** Source.js.orig	2020-07-27 17:32:45.835862380 +0000
--- Source.js	2020-07-28 17:27:21.323118426 +0000
***************
*** 26,33 ****
--- 26,37 ----
          Object.assign(this, options);
      }

      createNodes(nodes) {
+         const add = TmshUtil.addBulkNodes(nodes);
+         //return Promise.all([add]);
+         return add.then();
+         /*
          const adds = nodes.map(node => TmshUtil.addNode(node)
              .catch((error) => {
                  if (!error.message.includes('already exists')) throw error;
                  return TmshUtil.getNodeByIp(node.ip)
***************
*** 36,43 ****
--- 40,48 ----
                          node.id = conflict.id.charAt(0) === '/' ? conflict.id : `/${conflict.id}`;
                      });
              }));
          return Promise.all(adds);
+         */
      }

      deleteNodes(nodes) {
          const deletes = nodes.map(node => TmshUtil.deleteNode(node.id)
```

**Comment by delgadillo22** [comment]
Thank you -- tracking via AUTOTOOL-1751

**Comment by chen23** [comment]
I believe this should also address #189 and #188

**Comment by chen23** [comment]
Will also handle #287

**Comment by chen23** [comment]
Fixed in AS3 3.24
